### ISSUE TITLE
DefaultTask.inputs exposes internal API

### ISSUE THREAD

**Issue by cy6erGn0m** [gradle/gradle#1004]
### Expected Behavior
`DefaultTask.inputs` should have type `TaskInputs`

### Current Behavior
`DefaultTask.inputs` has type `TaskInputsInternal` of an internal package. So if there are changes in the internal class then user's code could corrupt. 

### Context
My own plugin fails with NoSuchMethodError with older Gradle because of unexpected internal API leakage

### Steps to Reproduce (for bugs)
1. Create plugin project, implement and add your own task (implementation should be written in Java) by deriving `DefaultTask`, in the constructor call `getInputs().file(project.buildFile)` (it could be any other file)

2. Compile and run your task under 3.2.1 - pass
3. Use compiled plugin in gradle 3.1 - fail with NoSuchMethodError

### Your Environment
Gradle 3.2.1, Gradle 3.1

* Internal code review: PR #1021 (was: ~~[REVIEW-6409](https://code-review.gradle.org/cru/REVIEW-6409)~~)

**Comment by lptr** [comment]
`AbstractTask.getInputs()` returning the internal `TaskInputsInternal` was introduced in 3.0-milestone-1: https://github.com/gradle/gradle/commit/acdc275ed043ce41b619a9f4db53d7f01009c554

So the issue seems legit between Gradle 2.x and 3.x. What's strange to me is how it is possible to reproduce the problem between 3.2.1 and 3.1, as there was no change related to `Task.getInputs()` since 3.0. I'll investigate this later.

The problem impacts tasks that were compiled against Gradle 3.x that use `AbstractTask.getInputs()` directly. It does not impact tasks that use `getOutputs()` only (it has been returning an internal type since Gradle 0.9), or code that calls `Task.getInputs()` from outside the task (like plugins and build scripts). Tasks compiled against 2.x are in general unaffected, as `Task.getInputs()` provides the necessary bridge method with the `TaskInputs` return type in 3.x as well.

We can't do much about tasks compiled against Gradle 3.0..3.2.1 failing when executed with Gradle 2.x, but we can fix this for 3.3 and later. We could do the following:

1. revert `AbstractTask.getInputs()` to return `TaskInputs`, so that tasks compiled against 3.3 would work fine with 2.x,
2. use `MixInLegacyTypesClassLoader` to weave in a bridge method that returns `TaskInputsInternal`, so tasks compiled against 3.0..3.2.1 would work with 3.3+,
3. perhaps do the same with `getOutputs()` to get rid of the internal return type there as well.

**Comment by cy6erGn0m** [comment]
The difference between versions in `TaskInputsInternal`:
https://github.com/gradle/gradle/blob/v3.1.0/subprojects/core/src/main/java/org/gradle/api/internal/TaskInputsInternal.java

vs 
https://github.com/gradle/gradle/blob/v3.2.0/subprojects/core/src/main/java/org/gradle/api/internal/TaskInputsInternal.java

So if we have 3.2.0 classpath then the compiler emis invoke to `TaskInputsInternal.file()` that doesn't exist in 3.1.0

**Comment by lptr** [comment]
Thanks for the clarification @cy6erGn0m.

Here's a timeline of events:

* 0.9-preview-1 (38dc95cc95226d3b66004cca6f415a9c5a90ba07) –  `TaskOutputsInternal` is introduced, and `TaskInternal.getOutputs()` returns it instead of `TaskOutputs`. As `TaskOutputsInternal` doesn't override any public method, it's not causing any problems.
* 3.0-milestone-1 (acdc275ed043ce41b619a9f4db53d7f01009c554) – `TaskInputsInternal` is introduced
* 3.0-rc-1 (ecb2b083c7e56434cdf5eed437868e80e6edc39f) – `TaskInputsInternal.getFileProperties()` is added. As it overrides no public API, no problem is caused by this yet.
* 3.2 (683e60ac78ab49a627646aeabd25077d3d57929a) – `TaskInputsInternal.file()`, `files()` and `dir()` are added. Tasks compiled against this version will refer to `TaskInputFilePropertyBuilderInternal` as the return type of these methods.

So the concrete problems are caused by the overrides of these three public methods returning internal types. Removing the three overrides from `TaskInputsInternal` would break tasks that used them that were compiled against 3.2 and 3.2.1 with any other version, while all other versions would be still compatible with each other.

![image](https://cloud.githubusercontent.com/assets/495366/21173552/9658c9b0-c1d8-11e6-8f36-9798d583baaa.png)

**Comment by lptr** [comment]
For reference, here's a test project to reproduce the problem:

[taskinputs-internal-leak.zip](https://github.com/gradle/gradle/files/651044/taskinputs-internal-leak.zip)

Running `./gradlew help` in the root should produce:

```text
Caused by: java.lang.NoSuchMethodError: org.gradle.api.internal.TaskInputsInternal.file(Ljava/lang/Object;)Lorg/gradle/api/internal/tasks/TaskInputFilePropertyBuilderInternal;
        at CustomTask.<init>(CustomTask.java:6)
        at CustomTask_Decorated.<init>(Unknown Source)
        at org.gradle.api.internal.DependencyInjectingInstantiator.newInstance(DependencyInjectingInstantiator.java:56)
        at org.gradle.api.internal.ClassGeneratorBackedInstantiator.newInstance(ClassGeneratorBackedInstantiator.java:36)
        at org.gradle.api.internal.project.taskfactory.TaskFactory$1.call(TaskFactory.java:121)
        ... 75 more
```

**Comment by big-guy** [comment]
@cy6erGn0m Could you try taking a look at this with 3.3-rc-1?  Plugins built with 3.2/3.2.1 will be incompatible with other 3.x releases, but plugins built with 3.0/3.1/3.3 should be compatible with all 3.x releases.

**Comment by eriwen** [comment]
@big-guy @lptr Is this still in progress?

**Comment by lptr** [comment]
@eriwen: no, this was fixed in 3.3 by reverting to the behavior in 3.1.

**Comment by eriwen** [comment]
In that case I'll close this as delivered 3.3
